### PR TITLE
[refs #218] Create a changelog file

### PR DESCRIPTION
**Pull request by herzinger**
Create `CHANGELOG.md`, with a pattern to follow, and add instructions on how to update it to `CONTRIBUTING.md`

### CODE REVIEW

**Comment by florianbouvot** [comment]
Isn't it too much? I'm ok for adding a changelog but your version is like commit history, why not something lighter (and more readable)?

**Comment by herzinger** [comment]
What I did is pretty much the same as the one you linked, adding only the PR number and link (which I find REALLY useful, since with this way you can se exactly what was changed to adjust your code accordingly) and separation between possible breaking changes and other, safer modifications. I find those additions necessary, since inuitcss is considerably more complex than normalize.css. We have more than one file, to begin with.

>your version is like commit history, why not something lighter (and more readable)?

As I see it, this is really light and readable. I even edited the descriptions so they are easier to understand by themselves and grouped PRs with deal with the same (or similar) issues. Just as a comparison, this is the full commit history for the same period, as it's generated by [git-extras' git changelog](https://github.com/tj/git-extras/blob/master/Commands.md#git-changelog).

```markdown
6.0.0-beta.5 / YYYY-MM-DD
=========================

  * Merge pull request #216 from inuitcss/tkt-0216
  * [refs #216] Fix transition-duration value
  * Merge pull request #215 from inuitcss/tkt-0214
  * Use simpler value for example global transitions variable.
  * Merge pull request #213 from inuitcss/tkt-0212
  * [refs #212] Use clip-path as progressive enhancement
  * [refs #212] Fix shrinked text : preserve space between words
  * Merge pull request #211 from inuitcss/tkt-0209
  * Merge pull request #209 from inuitcss/tkt-0208
  * [refs #209] Rename tables objects : use singular
  * [refs #208] Fix object name in comment
  * Merge pull request #205 from inuitcss/tkt-0021
  * Merge branch 'develop' into tkt-0021
  * [refs #21] Remove now needless warning


6.0.0-beta.4 / 2016-10-18
=========================

  * v6.0.0-beta.4
  * Merge pull request #203 from inuitcss/tkt-0200
  * Merge pull request #193 from inuitcss/tkt-0156
  * Merge pull request #191 from inuitcss/tkt-0159
  * [refs #200] Be correct about !important
  * Merge pull request #202 from inuitcss/tkt-0199
  * [refs #200] Reset `left`/`right` property correctly
  * [refs #199] Remove Full Doc section
  * Merge pull request #194 from inuitcss/tkt-0058
  * Merge pull request #192 from inuitcss/tkt-0166
  * Merge branch 'tkt-0058' of github.com:inuitcss/inuitcss into tkt-0058
  * [refs #58] Add CircleCI badge
  * [refs #156] Remove TOC entry for forms partial
  * [refs #156] Transfer form resets to generic.reset.scss
  * [refs #159] Make spacing classes configurable
  * Merge pull request #185 from inuitcss/tkt-0162
  * Merge pull request #187 from inlikealion/tkt-0058
  * Fix spelling
  * [refs #58] Add section about inuitcss plugins
  * [refs #58] Add note about supported Sass versions
  * [refs #58] Add section about namespacing classes
  * [refs #58] Adopt core functionality section from pre-alpha
  * [refs #58] Add section about modifying inuitcss
  * [refs #58] Add section about extending inuitcss
  * [refs #58] Add responsive section to README
  * [refs #58] Add section about directory structure
  * [refs #58] Update listing of example files as per master branch
  * Merge branch 'develop' into tkt-0166
  * [refs #166] Provide widths-classes customization
  * Merge pull request #184 from inuitcss/tkt-0182
  * Merge pull request #176 from inuitcss/tkt-0116
  * [refs #116] Make rem mixin IE8-ready
  * Merge pull request #183 from herzinger/tkt-0181
  * [refs #162] Warn using px for font-size mixin
  * [refs #182] Update normalize.css 4.2.0 to 5.0.0
  * [refs #116] Update tests
  * [refs #116] Update stylelint config
  * [refs #116] Note about why we duplicate font-size declaration
  * [refs #116] Remove oldIE condition
  * [refs #116] Remove oldIE var
  * Merge branch 'develop' into tkt-0116
  * [refs #156] Add `elements.form` partial
  * [refs #116] Make it IE<9 compatible
  * [refs #116] Remove IE<9 support
  * Merge branch 'develop' into tkt-0058
  * [refs #0058] Fix typo
  * Merge branch 'develop' into tkt-0058
  * [refs #0058] Add instructions for example globals file
  * Merge branch 'develop' into tkt-0058
  * Merge branch 'develop' into tkt-0058
  * Merge branch 'develop' into tkt-0058
  * [refs #0058] Add basic instructions for components file
  * [refs #0058] Add basic instructions for config file
  * [refs #0058] Link file titles to make it easier for people to see what I’m on about
  * [refs #0058] Tidy up CLI instructions
  * [refs #0058] Add basic instructions for main file
  * [refs #0058] Use logo as README title
  * Merge branch 'develop' into tkt-0058
  * Crude installation docs
  * Merge branch 'develop' into tkt-0058
  * Merge branch 'develop' into tkt-0058
  * Merge branch 'develop' into tkt-0058
  * [refs #0058] Add note about updating file import paths
```

**Comment by florianbouvot** [comment]
I can be wrong but for me:

* changelog give a global view of project update
* commit history give full update details

So I think we need a basic changelog to help people without full history...

**Comment by herzinger** [comment]
Yes, I agree. And that's exactly what I provided.

**Comment by florianbouvot** [comment]
I notice that popular repo like bootstrap and foundation uses GitHub's releases feature instead 👍 (less work and same info).

**Comment by herzinger** [comment]
@florianbouvot releases are great. But then you guys will have to create the changelogs for each tag by yourselves at the time of each version bump. Believe me, creating a changelog from the commit history is a lot more work (it's the work I did to create this starting point). 

With this changelog, each contributor adds their own entry, in their own pull requests. I'd use both, so that you can just copy the appropriate section of this file and paste it in the release when creating it.

**Comment by florianbouvot** [comment]
@herzinger I understand your points, I will think about it this night and wait for @inuitcss/core opinion 😉

**Comment by herzinger** [comment]
More on that: releases and changelog are NOT the same thing. Each release contains a changelog for that specific release. This is a full changelog of the project, which is useful if you are updating in a jump between more than one version bump of difference. Also, this file is subjected to version control at every change, which is not the case for the mini changelogs used in releases. Last but not last, the changelogs in releases are NOT automatically generated, you need to type 'em in just the same, so maintaining this file is actually a lot easier (especially for the reason I pointed out above).

**Comment by anselmh** [comment]
Coming from various open source projects here, I have some insights. For all the options we have, there are advantages and disadvantages. Let me try to explain the differences here.

### Github Releases

* Easy way to write release notes via an UI during a release
* No fixed format (no templates possible for it, but since this is up to the core, it doesn’t matter)
* Easy to follow along since this is where people expect announcements, breaking changes, bugfixes, etc.
* Not platform independent (can’t be transferred to another service since it lives on Github, not git)

### Git Tags + Commit History

* Easy since nearly no extra effort is needed
* CLI-Tools / git hook tools can help writing a release tag commit message (including the same as you’d write into github releases)
* Commit history is easily cluttered
* Writing git tag commit messages is easily being forgotten during releases
* Independent of the platform since it only uses git’s native features

### CHANGELOG.md

* Easy to write
* lives in repository
* easy to discover for everybody
* hard to tell people to include meaningful changelog messages with their Pull Requests
* hard to tell people they need to update the file all the time before releases (and commit it during the git version update commit)

In summary, I have tried and used all of these methods in various projects. **The easiest and most robust way was using Github releases**.  
I think we should decide this as a team, my vote would go for Github releases.

@herzinger whatever we decide, I appreciate the work you’ve done here so far, since it also points out what maintaining a CHANGELOG.md file means. So a big thank you for suggesting this. :)

**Comment by herzinger** [comment]
@anselmh don't mention it. 

I would just like to clarify that I'm all for using releases. But as I said, one does not eliminate the other. Actually, all the options you listed should be used, to some degree.

* **Commits history** is an inherent part of git, so no discussion here.
* The real shine in **releases** appears when used in combination with **git tags**.
* Maintaining a **`CHANGELOG.md`** makes it easier to document releases, while bringing extra value by itself.

The extra value comes from:

1. Having in one, clear and easy to find, full log of all changes in the project, across versions.
2. Making new changes easier to track during pre-releases, since you are adding them along the way.
3. Having the logs as part of the version control.

>* hard to tell people to include meaningful changelog messages with their Pull Requests
* hard to tell people they need to update the file all the time before releases (and commit it during the git version update commit)

That's why I included it in our `CONTRIBUTING.md` file. We already follow and impose a series of unusual guidelines, I don't think this extra one should be a problem. Enforcing the usage is easy, since all contributions come from pull requests that, obligatorily, need to be approved before merge. Even if the core team choses to add the entries themselves, it's still easier to update somewhere along the way, at the time the changes are make, ratter than only at the time of the version bumps.

**Comment by anselmh** [comment]
I just wonder if maintaining a git commit history, releases AND a changelog.md file is worth the extra effort. Also, I don’t see where a release should differ from a changelog, except that changelog.md has a section for unreleased changes, but apart from that it should be written the same way to make any sense.

**Comment by herzinger** [comment]
@anselmh Yes, but each release documents only the changes that are part of that particular release, while the changelog file has all the changes in the project in one place. Imagine someone updating from beta.1 to beta.4, for example. Also, there is no extra work, since at the release time you would just copy/paste the relevant section from the full changelog. It may seen counter intuitive, but maintaining both makes it LESS work when it comes the time to make an new bump. Also, as you can see in [this link](https://github.com/blog/1547-release-your-software) (that I also added in my previous answer), it's possible to create tags and releases with a single action, so very little would change from how it's done today.

**Comment by anselmh** [comment]
Huh, I don’t get your arguments now anymore, I’m sorry. I’ve done a lot of releases already, so yes I know how to do this in the most efficient way.

> Also, there is no extra work, since at the release time you would just copy/paste the relevant section from the full changelog. It may seen counter intuitive, but maintaining both makes it LESS work when it comes the time to make a new bump. 

This is simply wrong. Editing CHANGELOG.md to contain the proper new release version ahead of the specific commits, committing this, copying text from the CHANGELOG.md to the github release are two additional steps that need to be done for a release.
The only benefit might be that the changes are already written down in a user-readable way. However, this is not something I see as a huge blocker for doing a release. You need to review a release anyway again for all its changes and check if it works before publishing, which is why I don’t see a reason why it would _save_ time to add a manually written and edited CHANGELOG.md file.

**Comment by herzinger** [comment]
@anselmh because writing down the changes in a readable way would become part of the contributing process. That means it wouldn't be one single person's job, and it would not be done all at a single time. Searching through all the history whiting a release, filtering what is or is not relevant, rewriting descriptions and squashing similarities or redundancies to make a concise changelog is a lot of work, especially when you have different people contributing. It's a lot easier to just review it, and less prone to errors.

I did my part, and raise all the arguments, so I'll let you guys decide on how to go from here. I'd be glad to maintain it all if I had the access to do so. From here on, I'll be fine with whatever decision is made, as long as in the end we have a better documentation of changes than what we have today (none, apart from what is unavoidable when using a version control).

**Comment by herzinger** [comment]
Oh, just a little something extra: maintaining a changelog.md in parallel let's you use github releases while eliminating all of the disadvantages @anselmh listed here:

>* No fixed format (no templates possible for it, but since this is up to the core, it doesn’t matter)
* Not platform independent (can’t be transferred to another service since it lives on Github, not git)

**Comment by csshugs** [comment]
cc @inuitcss/core 

> [...] writing down the changes in a readable way would become part of the contributing process.

@herzinger you convinced me here. 
This is the first open source project I'm involved in, so I might not have the experience in which way of documenting this is the best and what the pros and cons are.

When I understand **@herzingers** point correctly, it would mean that with every PR, the changes made in this PR have to be documented in `CHANGELOG.md`. So extending the `CHANGELOG.md` with that information is part of the PR. In other words: In every PR, there must be changes in the `CHANGELOG.md` file. Otherwise, the PR cannot get approved. I think, this is a valid convention. And since every PR needs to be approved anyway, in case the person who opened the PR forgot to mark the changes, other contributors/core member can indicate that something is missing. 
I'm with @herzinger: This is a little bit extra work to do per PR.

**But:** When it's time for a new release, the inuitcss core member publishing the release doesn't have to put a lot of effort in investigating what has been changed since the last release—It's all already there in `CHANGELOG.md`. It's more or less just copy and pasting and formatting the release note. At this point we would benefit from making it a part of the contribution process of marking out each change in the respective PR, instead of collecting all this information afterwards when releasing a new version.

In summary: I'm in favor of approving this PR, thus providing a `CHANGELOG.md` which is maintained alongside with every code change. On top of that, we should use GitHub Releases, but maybe that can wait until we're out of beta.

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+==================
+
+## Possible breaking changes
+  * [[PR #211]](https://github.com/inuitcss/inuitcss/pull/211) - Rename tables objects: use singular
+
+## Other fixes or features
+  * [[PR #229]](https://github.com/inuitcss/inuitcss/pull/229) - Create this changelog file
+  * [[PR #215]](https://github.com/inuitcss/inuitcss/pull/215) and [[PR #216]](https://github.com/inuitcss/inuitcss/pull/216) - Use simpler value for example global transitions variable
+  * [[PR #213]](https://github.com/inuitcss/inuitcss/pull/213) - add some rules to the `inuit-hidden-visually` mixin
+  * [[PR #209]](https://github.com/inuitcss/inuitcss/pull/209) - Fix typo in the title banner in `objects.wrapper`
```
Comment (herzinger) [inline CHANGELOG.md]: Hum... I think the line is a bit too blurry to make a good distinction, but I can give it a shot if you guys think the added value is worth it.

Comment (herzinger) [inline CHANGELOG.md]: I do like a lot the idea of giving special focus on new stuff.

Comment (herzinger) [inline CHANGELOG.md]: @csshugs Take a look now. I even added some marketing spice hahaha

**Comment by florianbouvot** [comment]
@csshugs I think a "minimal" `CHANGELOG` might be suitable, but because I haven't experience I can be wrong... So I agree about `CHANGELOG` but I wonder if we need to start now or at 6.0.0?

**Comment by herzinger** [comment]
@florianbouvot is there any down side to starting now? Especially since this here is ready?

**Comment by csshugs** [comment]
@herzinger @florianbouvot 

> is there any down side to starting now? Especially since this here is ready?

Agree. Since this is ready to use already, It just cannot hurt to integrate it now. The are already people using the beta after all.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
+
+6.0.0-beta.5 / YYYY-MM-DD
```
Comment (anselmh) [inline CHANGELOG.md]: I have concerns using unreleased version numbers in CHANGELOG. This can easily lead to faulty version numbers, missing tags, or changes not being released but in the changelog. Especially when you think of a bugfix release, this usually shouldn’t include new features. I have seen too many faulty release workflows that I can approve this as good practice, sorry. I’d like to have all the unreleased changes under the title `HEAD`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+==================
+
+## New features
```
Comment (anselmh) [inline CHANGELOG.md]: If we’re going to have these categories/subtitles, we also need to differentiate bugfixes from improvements. I also _personally_ don’t like the "Other, safer changes" since it doesn’t say what this is—is it a removal of something, a new feature, an improvement?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -68,6 +68,12 @@ follow work through the issue/pull request process.
 5. **Open a pull request into `develop`.** Our main branch is `develop`, and
    this is where we want new work to be merged into. `master` is a release
    branch only.
+6. **Update `CHANGELOG.md` in your branch.** Add the pull-request number and 
+   link, and a brief, one line description under the next tag title. Make sure
```
Comment (anselmh) [inline CONTRIBUTING.md]: Adding a PR’s number/link means an author has to create a pull request, then add a new commit again and push it again. This is very confusing, especially for beginners and also messes up with the git history since the changelog notes won’t belong to the actual commit of the change. Also, it introduces the work for maintainers since it remains first unclear if a Pull Request is ready to review or not and if someone forgot to link to its own PR, we need to wait a bit and then ask again if this has been forgotten.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -68,6 +68,12 @@ follow work through the issue/pull request process.
 5. **Open a pull request into `develop`.** Our main branch is `develop`, and
    this is where we want new work to be merged into. `master` is a release
    branch only.
+6. **Update `CHANGELOG.md` in your branch.** Add the pull-request number and 
+   link, and a brief, one line description under the next tag title. Make sure 
+   the reason for the change can be understood. If this change has a chance of 
+   making an existent project's build fail, or cause unexpected changes in the 
+   output, group it under the "Possible breaking changes" title. New features
```
Comment (anselmh) [inline CONTRIBUTING.md]: I don’t think this makes a clear statement on what a breaking change is. Basically, anything that would alter the visual output of an existing project (not including custom extensions) after an update and isn’t a bugfix is a breaking change. It also should be made clear that due to semver this means the need for a new major release. I think that is an important fact to add since many people don’t think about this beforehand and just go by "this is a little break but it’s easy to fix for users", which is a fallacy that we won’t go here at inuitcss. We want to provide a well working toolset that is maintained in a solid way and does follow semver as it’s intended.

**Comment by anselmh** [comment]
> In summary: I'm in favor of approving this PR, thus providing a CHANGELOG.md which is maintained alongside with every code change. On top of that, we should use GitHub Releases, but maybe that can wait until we're out of beta.

I disagree here. I think we should use github releases for all our releases and right from the start now. Since we’re publishing official packages already to bower and npm, it makes sense to "officially" publish them as release in github as well.

**Comment by csshugs** [comment]
Ok, using GitHub releases right from the start is fine for me.

**Comment by herzinger** [comment]
I'm fine with everything here. I don't have the time to update it right now, but I'll try and address every concern until the weekend, as soon as I can.

**Comment by anselmh** [comment]
Cool, and no hurry @herzinger!

File: CONTRIBUTING.md (hunk 4 above)
Comment (herzinger) [inline CONTRIBUTING.md]: I still think it should be added. Without it we would need to be a lot more descriptive about each change, specially breaking changes. With the PR link it becomes easy to look up exactly which lines in which files were changed. Without it the only indication about what was done and how to adapt to it would have to be de description, and I would like to keep it short, one liners.

>[...] the changelog notes won’t belong to the actual commit of the change

I don't think they should. I would, one way or another, add it as a dedicated commit, the last on the PR, like: `[refs #xxxx] add changelog entry`. Also, it's pretty uncommon for a PR to be reviewed right after it's submitted, so I don't think that little delay would be a problem. You are right about everything else, but I think it's worth it, no? Or is there a easier way to point the reader to the exact changes, in a easy to visualize way?

File: CHANGELOG.md (hunk 3 above)
Comment (herzinger) [inline CHANGELOG.md]: My idea was to have 3 practical categories:

1. "Here are some new stuff you should check out"
2. "Here are some stuff you need to be aware of"
3. "Here are some stuff you don't need to worry about"

But I'll try and divide it up the way you suggested.

File: CONTRIBUTING.md (hunk 4 above)
Comment (florianbouvot) [inline CONTRIBUTING.md]: That's why I use Git history 😆 (I check Git release note for basic info and Git history for details).

Comment (herzinger) [inline CONTRIBUTING.md]: Yeah, but the commits history is not really user friendly. I see commits more as a tool for the developers than a tool for the users. The `files changed` tab in the pull-request page is the best way to look, in one single place, at all the code in a feature/fix/change. Also, it becomes easy to see all the discussion involved in that change, and it's a nice, subtle way to give credit to the PR author/contributors, since their names are displayed there.

**Comment by csshugs** [comment]
@herzinger 
Don't wanna rush you, but do you still plan to tackle the open points @anselmh requested?

**Comment by herzinger** [comment]
@csshugs I do. I'm in a vacation in the woods right now, but I'll get back to it right after the holidays.

**Comment by herzinger** [comment]
This is long overdue, sorry about that. I'll be making the changes and updating it this weekend. 

I think that if the currently open PR's are merged, we can make a version bump and release, this one beeing the only one that still needs work, as I see it.

**Comment by KingScooty** [comment]
Hi all.

Any news on this PR? I'm conscious we've spent 6 months trying to figure out what the changelog should look like, when there are features and changes being merged into `develop` that could be released.

**Comment by herzinger** [comment]
Yeah, sorry... That's the biggest hurdle to be honest. Making the changes asked is no big deal, I just need to find the time to actually update it with everything that has being done since, which is the work it will make easier once this is up. I'm a little busy lately with other things, but I'll try and make it soon

**Comment by csshugs** [comment]
Closing this since #309 is merged. 

Thanks anyway @herzinger for your great effort and contribution to this topic!
